### ISSUE TITLE
Scaffold for typescript in `@wordpress/create-block`

### ISSUE THREAD

**Issue by AdsonCicilioti** [WordPress/gutenberg#39915]
## What problem does this address?
For those who already use Typescript, it is extremely unsatisfactory to have to work with JS. Even more so when getting to know a new library and its API.

## What is your proposed solution?
The `@wordpress/create-block` package could have a scaffold for Typescript. This would help a lot in learning and greatly improve the developer experience.

**Comment by gziolo** [comment]
@ryanwelcher, do you have a template for Create Block that works with TypeScript?

**Comment by Mamaduka** [comment]
An excellent idea for the "Create Block" template.

I'm going to remove the "Needs Testing" label.

**Comment by ryanwelcher** [comment]
This doesn't need to be part of the tool but rather can be an external template. The two templates that are included are the most common use cases and I wonder if enough block developers are using Typescript to make this part of the tool instead of an external template. 

To be clear, I LOVE the idea of having a template that supports TS and this is where we should be headed. 

However, in my ( admittedly limited ) experience working with blocks and TS, the main pain point is the lack of available types. There has been an effort to add types to many of the packages in GB but not for some of the packages that block developers will use, such as the `@wordpress/blocks` package and the types available at [DefinitelyTyped ](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/wordpress__blocks) don't reflect the current APIs. There is an effort to move those into the project ( https://github.com/WordPress/gutenberg/pull/43686 ) but it has stalled - which is on me.

I think that before we move a full TS template into the tool, we should get the types in place for the packages that block devs will use to avoid the hassle of having to extend an out-of-date type set or having to roll their own types for built-in items.

This is a larger question that perhaps needs its own issue but perhaps we should consider creating this template as an external package in the repo. This would allow the types to be introduced in the template and then moved over to the corresponding packages as needed.

**Comment by gziolo** [comment]
> This is a larger question that perhaps needs its own issue but perhaps we should consider creating this template as an external package in the repo. This would allow the types to be introduced in the template and then moved over to the corresponding packages as needed.

I fully agree that the first step would be to create the external template hosted on npm or GitHub (assuming Create Block can consume GitHub repository) that works with TypeScript. I don't think we are in good shape in terms of types to offer an official template that is 100% TypeScript ready. However, having an unofficial template ready for the community would help find what's missing and bring more attention to tasks that eventually bridge the gaps.

**Comment by tresorama** [comment]
I would be good to have a semi-official scaffolded template in TS.

The good news about current situation is that webpack config that ships with `@wordpress/create-block` is already compatible with TS, it's necessary only to change extension of files to `.ts` and everyting compiles.

*Maybe is required also to create a tsconfig.json in your workspace, i don't remember exactly. But you can grap a prefilled json config from the `tsc` cli or from https://github.com/tsconfig/bases*

**Comment by tuanfront-end** [comment]
Hello @ryanwelcher,  hope there are new updates on this?

**Comment by luisherranz** [comment]
Just as a heads up, I was planning to include a new template to show how to use TypeScript with the Interactivity API in this pull request:

- https://github.com/WordPress/gutenberg/pull/64577

It might make sense to create a general TypeScript template where the block is interactive and the Interactivity API is also typed instead of creating a template just for the Interactivity API, but for the conversation here it seems like that's not something that will happen soon, is it? Should I continue with the Interactivity API-only template? What do you think?

**Comment by gziolo** [comment]
> Just as a heads up, I was planning to include a new template to show how to use TypeScript with the Interactivity API in this pull request:
>
> https://github.com/WordPress/gutenberg/pull/64577

I responded with the following https://github.com/WordPress/gutenberg/pull/64577#issuecomment-2306373037. The gist of it is, the main differences when using TypeScript is the  file extensions and types added to JavaScript code so there is no need for creating new templates. Instead, we can offer variants of existing templates using concept of [variants](https://github.com/WordPress/gutenberg/blob/trunk/packages/create-block/docs/external-template.md#variants).

**Comment by gziolo** [comment]
It looks like one of the prerequisites for adding full TypeScript support to Create Block would be adding first-class TypeScript support for the most important core packages. @manzoorwanijk is coordinating adding support to `@wordpress/block-editor` and `@wordpress/blocks` which are the biggest packages necessary to develop blocks:
- https://github.com/WordPress/gutenberg/issues/67691

**Comment by ryanwelcher** [comment]
I've recently been testing Typescript out with `@wordpress/scripts` and it's trivial to get it working. `typescript` is already a dependency and it would only take adding a `tsconfig.json` to the project and a `tsc` command to `package.json`.

After that is where things start to get complicated. There currently no reliable types available for `registerBlockType` and the only way to get it working is to define all of the items defined in `block.json` in the call as well - which is not ideal for a number of reasons.

I would say that before we can create a template/variant for TS, we need to get the types in place for `@wordpress/blocks` which would give extenders what they need for the basic scaffold. We'd need to fill in gaps from there.

In the meantime, defining  the contents of `tsconfig.json` is a discussion that we would have here. Currently, I am using the following which seems to a sane set of defaults for using TS to lint.
```json
{
	"compilerOptions": {
		
		"esModuleInterop": true,
		"skipLibCheck": true,
		"target": "es2022",
		"allowJs": true,
		"resolveJsonModule": true,
		"moduleDetection": "force",
		"isolatedModules": true,
		"verbatimModuleSyntax": true,

		
		"strict": true,
		"noUncheckedIndexedAccess": true,
		"noImplicitOverride": true,

		"jsx": "preserve",

		"module": "preserve",
		"noEmit": true,

		"lib": ["es2022", "dom", "dom.iterable"]
	},
	"include": ["src/**/*", "src/**/*.json"]
}
```

My TS knowledge is not as deep as I'd like so I'm happy to discuss/learn about these or other configs that may make sense.
